### PR TITLE
Use zarr v3 dimension_names

### LINKED ISSUE #9668: Zarr V3 support: use `dimension_names` array metadata rather than user attribute `_ARRAY_DIMENSIONS`
### What is your issue?

Zarr V3 now explicitly supports dimension names on arrays (see [Array API docs](https://zarr.readthedocs.io/en/v3/_autoapi/zarr/index.html#zarr.Array).

This makes our `_ARRAY_DIMENSIONS` special attribute unnecessary.
https://github.com/pydata/xarray/blob/b133fdca71fd614b8056c3e995062b2834320ba5/xarray/backends/zarr.py#L54-L55

However, we are still using `_ARRAY_DIMENSIONS` with V3 in Xarray.

It would be great to fix this before we release V3 support so we don't start putting V3 data out into wild with this attribute set.

### PR DESCRIPTION
**Pull request by rabernat**
- [x] Closes #9668 
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- ~New functions/methods are listed in `api.rst`~

### CODE REVIEW

File: xarray/backends/zarr.py [hunk 1]
```diff
@@ -308,6 +308,17 @@ def _determine_zarr_chunks(
...
+    # Zarr V3 explicitly stores the dimension names in the metadata
+    try:
+        # if this exists, we are looking at a Zarr V3 array
+        # convert None to empty tuple
+        dimensions = zarr_obj.metadata.dimension_names or ()
+        attributes = dict(zarr_obj.attrs)
+        return dimensions, attributes
+    except AttributeError:
+        # continue to old code path
+        pass
```
Comment (shoyer) [inline xarray/backends/zarr.py:320]: `AttributeError` is a pretty broad exception, best to make it as narrowly scoped as possible, e.g.,
```suggestion
    try:
        # if this exists, we are looking at a Zarr V3 array
        metadata = zarr_obj.metadata
    except AttributeError:
        # continue to old code path
        pass
    else:
        # convert None to empty tuple
        dimensions = metadata.dimension_names or ()
        attributes = dict(zarr_obj.attrs)
        return dimensions, attributes
```

**Comment by shoyer** [comment]
> Question: do we want to handle the case where V3 data have _ARRAY_DIMENSIONS attrs set?

Hopefully such Zarr v3 files do not exist! I am OK skipping this

**Review by shoyer — APPROVED** [review]

**Comment by shoyer** [comment]
Looks good, thanks Ryan!

At some point, it would also probably be a good idea to update our Zarr encoding docs: https://docs.xarray.dev/en/stable/internals/zarr-encoding-spec.html

**Review by dcherian — APPROVED** [review]

**Comment by TomNicholas** [comment]
Are the upstream errors expected / allowed for now?

**Comment by TomNicholas** [comment]
Actually [this error](https://github.com/pydata/xarray/actions/runs/11505105378/job/32026660292?pr=9669)

`AttributeError: 'ArrayV2Metadata' object has no attribute 'dimension_names'`

does seem like its relevant to this PR

**Comment by rabernat** [comment]
Yes thanks for flagging. I'm blaming @shoyer's suggestion in fc3d4c3280f7190719f8e0baf703b783715bcb3b, which broke compatibility with V2 metadata. 😛 Will push a fix later tonight.

File: xarray/backends/zarr.py [hunk 2]
```diff
@@ -308,6 +308,19 @@ def _determine_zarr_chunks(
...
+        # if this exists, we are looking at a Zarr V3 array
+        metadata = zarr_obj.metadata
+    except AttributeError:
+        # continue to old code path
+        pass
+    else:
+        # convert None to empty tuple
+        dimensions = metadata.dimension_names or ()
+        attributes = dict(zarr_obj.attrs)
+        return dimensions, attributes
```
Comment (jhamman) [inline xarray/backends/zarr.py:322]:
```suggestion
    try:
        # if this exists, we are looking at a Zarr V3 array
        # convert None to empty tuple
        dimensions = zarr_obj.metadata.dimension_names or ()
    except AttributeError:
        # continue to old code path
        pass
    else:
        attributes = dict(zarr_obj.attrs)
        return dimensions, attributes
```

Comment (TomNicholas) [reply xarray/backends/zarr.py:322]: I say just try adding this commit now

**Comment by TomNicholas** [comment]
This looks good now - the only remaining upstream failures are groupby-related. Merge?

**Pull request merged**